subsys/bluetooth/host/scan.c

```diff
@@ -765,154 +872,285 @@ static void create_ext_adv_info(struct bt_hci_evt_le_ext_advertising_info const
-		more_to_come = data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_PARTIAL;
+static void crash_if_corrupted(size_t report_length, size_t buffer_length)
+{
+	BT_ASSERT_MSG(buffer_length >= report_length, "Adv report corrupted (wants %u out of %u)",
+		      report_length, buffer_length);
+}
 
-		if (evt->length > buf->len) {
-			LOG_WRN("Adv report corrupted (wants %u out of %u)", evt->length, buf->len);
+enum data_status {
+	COMPLETE = 0x00,
+	INCOMPLETE_MORE_DATA_TO_COME = 0x01,
+	INCOMPLETE_DATA_TRUNCATED_NO_MORE_TO_COME = 0x02,
+	STATUS_RFU = 0x03,
+};
 
-			net_buf_reset(buf);
+extern struct k_work_q *bt_get_rx_wq(void);
 
-			if (evt_type & BT_HCI_LE_ADV_EVT_TYPE_LEGACY) {
-				return;
-			}
+static void process_reassembled_report_async(struct net_buf *report, bt_addr_le_t *addr,
+					     struct bt_le_scan_recv_info *info)
+{
+	net_buf_push_mem(report, info, sizeof(*info));
+	net_buf_push_mem(report, addr, sizeof(*addr));
+	net_buf_push_u8(report, REPORT_MAGIC_REASSEMBLED);
 
-			/* Start discarding irrespective of the `more_to_come` flag. We
-			 * assume we may have lost a partial adv report in the truncated
-			 * data.
-			 */
-			reassembling_advertiser.state = FRAG_ADV_DISCARDING;
+	LOG_DBG("queue %p", report);
 
-			return;
-		}
+	k_fifo_put(&adv_reports, report);
+	k_work_submit_to_queue(bt_get_rx_wq(), &adv_report_work);
+}
 
-		if (evt_type & BT_HCI_LE_ADV_EVT_TYPE_LEGACY) {
-			/* Legacy advertising reports are complete.
-			 * Create event immediately.
-			 */
-			create_ext_adv_info(evt, &scan_info);
-			le_adv_recv(&evt->addr, &scan_info, &buf->b, evt->length);
-			goto cont;
-		}
+static char *status_to_str(uint8_t status)
+{
+	switch (status) {
+	case COMPLETE:
+		return "COMPLETE";
+	case INCOMPLETE_MORE_DATA_TO_COME:
+		return "INCOMPLETE-MD";
+	case INCOMPLETE_DATA_TRUNCATED_NO_MORE_TO_COME:
+		return "INCOMPLETE-NM";
+	case STATUS_RFU:
+		return "RFU";
+	default:
+		__ASSERT(0, "Unknown data status %x", status);
+		return "";
+	}
+}
 
-		is_new_advertiser = reassembling_advertiser.state == FRAG_ADV_INACTIVE ||
-				    !fragmented_advertisers_equal(&reassembling_advertiser,
-								  &evt->addr, evt->sid);
+static bool process_fragmented_report(struct bt_hci_evt_le_ext_advertising_info *evt)
+{
+	/* This function parses the event and re-assembles fragmented reports
+	 * into synthetic events.
+	 *
+	 * Returns true if the event was processed. I.e. will return false for
+	 * reports that fit a single event and are complete.
+	 */
+	struct fragmented_advertiser *r = &reassembling_advertiser;
 
-		if (is_new_advertiser && is_report_complete) {
-			/* Only advertising report from this advertiser.
-			 * Create event immediately.
-			 */
-			create_ext_adv_info(evt, &scan_info);
-			le_adv_recv(&evt->addr, &scan_info, &buf->b, evt->length);
-			goto cont;
-		}
+	uint16_t evt_type = sys_le16_to_cpu(evt->evt_type);
+	uint16_t data_status = BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS(evt_type);
```

Comment:
`status_to_str` is called with `data_status` which has the return value of `BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS`. 

However the values returned by `BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS` are not the same values as `status_to_str` matches on. 

`BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS` should return one of the following: 

```c
BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE
BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_PARTIAL
BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_INCOMPLETE
```